### PR TITLE
[#32] Improve built-in /metrics endpoints to use both counter and gauge

### PR DESCRIPTION
**Pull request by resyfer**
@jesperpedersen This is the implementation of what I had in my mind for #32. I still haven't added an install instruction for `internal.yml` file to have the path `/etc/pgexporter/internal.yml`, but using the following command before running the binary should do it.

```sh
sudo cp ./contrib/yaml/internal.yml /etc/pgexporter/internal.yml
```

I am not much familiar with writing CMake files or Spec files for RPM, so I'll look into those for this when this idea gets a green signal.

PS: This is just a bare-bones working change to the repo to demonstrate my idea. This needs work here and there, which I'll be glad to do.

### CODE REVIEW

**Comment by jesperpedersen** [comment]
Ok, I get where you are coming from.

However, `internal.yml` def lives in `src/yaml` instead, and you can't assume it should loaded from `/etc/pgexporter` as the configuration can be overwritten by `-c`.

Start by investigation if it can be embedded into the shared library, and loaded from there. Or do it as an `internal.h` with the text string.

Latter is likely more acceptable...

As is, no

**Comment by jesperpedersen** [comment]
Much happier with that - go ahead

**Comment by resyfer** [comment]
@jesperpedersen PTAL

**Comment by resyfer** [comment]
I have yet to identify the metrics that are better represented by counter than a gauge and I'll start looking into it.

File: src/include/internal.h [hunk 1]
```diff
@@ -0,0 +1,243 @@
+#ifndef PGEXPORTER_INTERNAL_H
```
Comment (jesperpedersen) [inline src/include/internal.h:1]: Remember header

File: src/include/pgexporter.h [hunk 2]
```diff
@@ -61,12 +61,13 @@ extern "C" {
 #define MAX_USERNAME_LENGTH  128
 #define MAX_PASSWORD_LENGTH 1024
 
-#define MAX_PATH 1024
-#define MISC_LENGTH 128
-#define NUMBER_OF_SERVERS 64
-#define NUMBER_OF_USERS    64
-#define NUMBER_OF_ADMINS    8
-#define NUMBER_OF_METRICS 256
+#define MAX_PATH              1024
```
Comment (jesperpedersen) [inline src/include/pgexporter.h:64]: If you are doing this then right-align

File: src/include/pgexporter.h [hunk 3]
```diff
@@ -84,15 +85,16 @@ extern "C" {
...
 
 #define LABEL_TYPE      0
 #define COUNTER_TYPE    1
 #define GAUGE_TYPE      2
 #define HISTOGRAM_TYPE  3
 
-#define SORT_NAME  0
-#define SORT_DATA0 1
+#define SORT_NAME 0
+#define SORT_DATA 1
```
Comment (jesperpedersen) [inline src/include/pgexporter.h:97]: Lets keep the `SORT_DATA0` name - we may want to sort by multiple data columns in the future

File: src/libpgexporter/yaml_configuration.c [hunk 4]
```diff
@@ -118,14 +138,9 @@ pgexporter_read_yaml(struct prometheus* prometheus, char* filename, int* number_
...
 
-   file = fopen(filename, "r");
-   if (file == NULL)
-   {
-      pgexporter_log_error("pgexporter: fopen error %s", strerror(errno));
-      return 1;
-   }
    if (!yaml_parser_initialize(&parser))
    {
+      printf("Hi\n");
```
Comment (jesperpedersen) [inline src/libpgexporter/yaml_configuration.c:143]: Hello !

File: src/main.c [hunk 5]
```diff
@@ -27,18 +27,20 @@
...
 
 /* pgexporter */
-#include <pgexporter.h>
 #include <configuration.h>
-#include <yaml_configuration.h>
+#include <internal.h>
 #include <logging.h>
 #include <management.h>
 #include <network.h>
+#include <pgexporter.h>
```
Comment (jesperpedersen) [inline src/main.c:35]: This one is always first

File: src/main.c [hunk 6]
```diff
@@ -27,18 +27,20 @@
...
 #include <prometheus.h>
 #include <queries.h>
 #include <remote.h>
 #include <security.h>
+#include <stdbool.h>
```
Comment (jesperpedersen) [inline src/main.c:40]: System include

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
Sounds good - I gave the patch a quick scan...

**Comment by resyfer** [comment]
@jesperpedersen I've made the requested changes, and I'll give a PTAL after finishing looking into which metrics need to be collectors

**Comment by jesperpedersen** [comment]
Merged.

Thanks for your contribution !

**Comment by resyfer** [comment]
@jesperpedersen Thank you for the review and thank you for the (very long) discussion around it! I hope this solves the problem that the issue aimed to point out.

**Pull request merged**